### PR TITLE
Reduce cold boot times by loading content and media only once on cold boot

### LINKED ISSUE #9779: Content and media loads twice from SQL to nucache on a fresh server
Media and Content loads twice into NuCache from the sql database when a fresh server (no nucache .db files and no last synced file). 
1st load is when the PublishedSnapshotService starts. Then once the DatabaseServerMessenger starts, the server is determined as out of sync, then the content and media is loaded from sql again. This significantly increases the time for a new instance to be ready to serve requests.

## Umbraco version

I am seeing this issue on Umbraco version: 8.11.1

Reproduction
------------

### Bug summary

Content and media is loaded twice from sql on startup when the app_data/TEMP directory is empty

### Specifics
```

{"@t":"2021-02-09T02:22:30.4893268Z","@mt":"{StartMessage} [Timing {TimingId}]","StartMessage":"Loading content from database","TimingId":"fdee707","SourceContext":"Umbraco.Web.PublishedCache.NuCache.PublishedSnapshotService","ProcessId":8744,"ProcessName":"iisexpress","ThreadId":1,"AppDomainId":2,"AppDomainAppId":"LMW3SVC13ROOT","MachineName":"","Log4NetLevel":"INFO ","HttpRequestNumber":1,"HttpRequestId":"424ac081-8206-4296-843b-85d2a2c25d43"}
{"@t":"2021-02-09T02:28:53.0727284Z","@mt":"{EndMessage} ({Duration}ms) [Timing {TimingId}]","EndMessage":"Completed.","Duration":382583,"TimingId":"fdee707","SourceContext":"Umbraco.Web.PublishedCache.NuCache.PublishedSnapshotService","ProcessId":8744,"ProcessName":"iisexpress","ThreadId":1,"AppDomainId":2,"AppDomainAppId":"LMW3SVC13ROOT","MachineName":","Log4NetLevel":"INFO ","HttpRequestNumber":1,"HttpRequestId":"424ac081-8206-4296-843b-85d2a2c25d43"}
{"@t":"2021-02-09T02:29:46.3024482Z","@mt":"{StartMessage} [Timing {TimingId}]","StartMessage":"Loading media from database","TimingId":"431c4fe","SourceContext":"Umbraco.Web.PublishedCache.NuCache.PublishedSnapshotService","ProcessId":8744,"ProcessName":"iisexpress","ThreadId":1,"AppDomainId":2,"AppDomainAppId":"LMW3SVC13ROOT","MachineName":"","Log4NetLevel":"INFO ","HttpRequestNumber":1,"HttpRequestId":"424ac081-8206-4296-843b-85d2a2c25d43"}
{"@t":"2021-02-09T02:34:10.7644523Z","@mt":"{EndMessage} ({Duration}ms) [Timing {TimingId}]","EndMessage":"Completed.","Duration":264461,"TimingId":"431c4fe","SourceContext":"Umbraco.Web.PublishedCache.NuCache.PublishedSnapshotService","ProcessId":8744,"ProcessName":"iisexpress","ThreadId":1,"AppDomainId":2,"AppDomainAppId":"LMW3SVC13ROOT","MachineName":"","Log4NetLevel":"INFO ","HttpRequestNumber":1,"HttpRequestId":"424ac081-8206-4296-843b-85d2a2c25d43"}
{"@t":"2021-02-09T02:34:16.7334473Z","@mt":"Initializing Umbraco internal event handlers for cache refreshing.","SourceContext":"Umbraco.Web.Cache.DistributedCacheBinderComponent","ProcessId":8744,"ProcessName":"iisexpress","ThreadId":1,"AppDomainId":2,"AppDomainAppId":"LMW3SVC13ROOT","MachineName":"","Log4NetLevel":"INFO ","HttpRequestNumber":1,"HttpRequestId":"424ac081-8206-4296-843b-85d2a2c25d43"}
{"@t":"2021-02-09T02:34:16.8644506Z","@mt":"Adding examine event handlers for {RegisteredIndexers} index providers.","RegisteredIndexers":3,"SourceContext":"Umbraco.Web.Search.ExamineComponent","ProcessId":8744,"ProcessName":"iisexpress","ThreadId":1,"AppDomainId":2,"AppDomainAppId":"LMW3SVC13ROOT","MachineName":"","Log4NetLevel":"INFO ","HttpRequestNumber":1,"HttpRequestId":"424ac081-8206-4296-843b-85d2a2c25d43"}
{"@t":"2021-02-09T02:34:16.9094477Z","@mt":"No last synced Id found, this generally means this is a new server/install. The server will build its caches and indexes, and then adjust its last synced Id to the latest found in the database and maintain cache updates based on that Id.","@l":"Warning","SourceContext":"Umbraco.Core.Sync.DatabaseServerMessenger","ProcessId":8744,"ProcessName":"iisexpress","ThreadId":1,"AppDomainId":2,"AppDomainAppId":"LMW3SVC13ROOT","MachineName":"","Log4NetLevel":"WARN ","HttpRequestNumber":1,"HttpRequestId":"424ac081-8206-4296-843b-85d2a2c25d43"}
{"@t":"2021-02-09T02:34:16.9674502Z","@mt":"{StartMessage} [Timing {TimingId}]","StartMessage":"Loading content from database","TimingId":"ff5eaac","SourceContext":"Umbraco.Web.PublishedCache.NuCache.PublishedSnapshotService","ProcessId":8744,"ProcessName":"iisexpress","ThreadId":1,"AppDomainId":2,"AppDomainAppId":"LMW3SVC13ROOT","MachineName":"","Log4NetLevel":"INFO ","HttpRequestNumber":1,"HttpRequestId":"424ac081-8206-4296-843b-85d2a2c25d43"}
{"@t":"2021-02-09T02:41:24.7408795Z","@mt":"{EndMessage} ({Duration}ms) [Timing {TimingId}]","EndMessage":"Completed.","Duration":427773,"TimingId":"ff5eaac","SourceContext":"Umbraco.Web.PublishedCache.NuCache.PublishedSnapshotService","ProcessId":8744,"ProcessName":"iisexpress","ThreadId":1,"AppDomainId":2,"AppDomainAppId":"LMW3SVC13ROOT","MachineName":"","Log4NetLevel":"INFO ","HttpRequestNumber":1,"HttpRequestId":"424ac081-8206-4296-843b-85d2a2c25d43"}
{"@t":"2021-02-09T02:41:24.8038806Z","@mt":"{StartMessage} [Timing {TimingId}]","StartMessage":"Loading media from database","TimingId":"4fe97e6","SourceContext":"Umbraco.Web.PublishedCache.NuCache.PublishedSnapshotService","ProcessId":8744,"ProcessName":"iisexpress","ThreadId":1,"AppDomainId":2,"AppDomainAppId":"LMW3SVC13ROOT","MachineName":"","Log4NetLevel":"INFO ","HttpRequestNumber":1,"HttpRequestId":"424ac081-8206-4296-843b-85d2a2c25d43"}
{"@t":"2021-02-09T02:46:05.0505265Z","@mt":"{EndMessage} ({Duration}ms) [Timing {TimingId}]","EndMessage":"Completed.","Duration":280246,"TimingId":"4fe97e6","SourceContext":"Umbraco.Web.PublishedCache.NuCache.PublishedSnapshotService","ProcessId":8744,"ProcessName":"iisexpress","ThreadId":1,"AppDomainId":2,"AppDomainAppId":"LMW3SVC13ROOT","MachineName":"","Log4NetLevel":"INFO ","HttpRequestNumber":1,"HttpRequestId":"424ac081-8206-4296-843b-85d2a2c25d43"}
```

### Steps to reproduce

Delete contents of App_Data/TEMP.
Start site.
Watch logs for loading content/media from the database entries.

### Expected result

Content and media is loaded only once on start up from SQL into nucache

### Actual result

Content and media is loaded twice on start up from SQL into nucache

### PR DESCRIPTION
**Pull request by nzdev**
### Prerequisites

- [x] I have added steps to test this contribution in the description below

If there's an existing issue for this PR then this fixes #9779

### Description
What?

Lets the PublishedSnapshotService know if the servermessenger is going to cold boot so PublishedSnapshotService does not attempt to load all content and media twice.

Why?
Load content and media only once on cold boot saving significant boot time on sites with large amounts of content/media . Saves on memory as otherwise there is 2 generations of every content and media in memory at boot.

How to test?
Delete the NuCache.db and server sync state text file. The log message about loading content from the database should appear only once.

### CODE REVIEW

**Comment by nzdev** [comment]
Seems to work well, halves the boot time.

**Comment by Shazwazza** [comment]
Hi @nzdev I can see this is merged to the nucache perf branch and just having a look.

Without this change, I presume what is happening is that we load in content/media nucache via the db files (if they are there) but then a cold boot is detected and we re-update nucache based on what is in the database? (just trying to clarify exactly what's going on).

In netcore we'll be able to fix this better, a large part of the issue is PublishedSnapshotService doing stuff in it's ctor.

**Comment by nzdev** [comment]
Original logic:
On a cold boot (no db files, no sync .txt)

1. PublishedSnapshotService loads the all the published content then all the media.
2. BatchedServerMessenger detects there is no sync state text file.
3. BatchedServerMessenger performs a refresh and forces PublishedSnapshotService  to load all the published content and media again.

New Logic:
On a cold boot (no db files, no sync .txt)

1. PublishedSnapshotService Asks the ISyncBootStateAccessor if Umbraco is cold booting, which returns that Umbraco is cold booting, so PublishedSnapshotService  does not load any media or content.
2. BatchedServerMessenger detects there is no sync state text file.
3. BatchedServerMessenger performs a refresh and forces PublishedSnapshotService  to load all the published content and

**Comment by Shazwazza** [comment]
Yep thanks. In the case where there are DB files but a cold boot occurs, it's still loading nucache twice, but first from the db files and then from the database. 👍

File: src/Umbraco.Web/PublishedCache/NuCache/PublishedSnapshotService.cs [hunk 1]
```diff
@@ -217,7 +223,12 @@ private void LoadCachesOnStartup()
         {
             var okContent = false;
             var okMedia = false;
-
+            if (_syncBootStateAccessor.GetSyncBootState() == SyncBootState.ColdBoot)
+            {
+                _logger.Warn<PublishedSnapshotService>("Sync Service is in a Cold Boot state. Skip LoadCachesOnStartup as the Sync Service will trigger a full reload");
```
Comment (Shazwazza) [inline src/Umbraco.Web/PublishedCache/NuCache/PublishedSnapshotService.cs:228]: This should just be info, else we have 2x warnings for every cold boot when cold boots are actually 'normal'

**Review by Shazwazza — APPROVED** [review]
I've tested these changes in another branch and all works as expected. We'll have to change this behavior in netcore but that's ok and I'll make sure that's done. @nul800sebastiaan or @emma-hq I'll tag as community-review, would be good to get some additional testing done and then merge. Cheers!

**Comment by Shazwazza** [comment]
Turns out I need parts of this for #8893

I've created a new branch for that and merged this in and will tweak from there. Updated PR coming soon.

**Pull request merged**